ci: compile Clang and compiler-rt in msan jobs

fanquake opened the pull request:
This reworks the MSAN CIs, to first compile Clang and compiler-rt (using GCC 12), and then, compile an MSAN instrumented libc++ using the just-built Clang 16. This fixes the `native_fuzz_with_msan` job, working around https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1005341, by not using the Debian provided Clang/LLVM.

Also included are changes to streamline how we use our "custom libc++", according to upstream: https://releases.llvm.org/16.0.0/projects/libcxx/docs/UsingLibcxx.html#using-a-custom-built-libc, as well as other minor cleanups in the CI configs.

An example job is currently running in the qa-assets repo: https://github.com/bitcoin-core/qa-assets/pull/129 (https://cirrus-ci.com/task/4632561431871488).

fanquake committed:
ci: standardize custom libc++ usage in MSAN jobs

Use `-isystem` & `-nostd*` flags, which is the preferred way to use a
custom libc++ (ours is libc++ build with MSAN) with Clang, as opposed to
our current ad-hoc flags.

See: https://releases.llvm.org/16.0.0/projects/libcxx/docs/UsingLibcxx.html#using-a-custom-built-libc
for more info.

fanquake committed:
ci: remove extra CC & CXX from MSAN jobs

This is passed through from depends.

fanquake committed:
ci: use LLVM 16.0.4 in MSAN jobs

fanquake committed:
ci: compile clang and compiler-rt in MSAN jobs

This works around https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1005341.

fanquake commented:
> No idea about the bash error. Maybe set -ex is missing in the file?

Have a change to just remove `cd` usage entirely, that should make the linter happy

maflcko reviewed (commented):
Do you happen to know what the difference to the debian package is? Maybe it can be fixed upstream instead?

fanquake committed:
ci: return to using Ubuntu 22.04 in MSAN jobs

We no-longer need to use 23.04, now that we aren't installing clang-16
and friends.

fanquake commented:
> Do you happen to know what the difference to the debian package is? Maybe it can be fixed upstream instead?

I'm not completely sure, the answer will be in here somewhere: https://salsa.debian.org/pkg-llvm-team/llvm-toolchain/-/blob/snapshot/debian/rules (also taking into account [all of the patches](https://salsa.debian.org/pkg-llvm-team/llvm-toolchain/-/tree/snapshot/debian/patches) they apply). They should be doing a multi stage build, and building everything with the 2nd-built Clang, but clearly something has been broken since LLVM 13. I'll followup upstream.

> Have a change to just remove cd usage entirely, that should make the linter happy

Made this change, and switched back to using Jammy. A qa-assets run of the current branch in: https://cirrus-ci.com/task/4632561431871488.

maflcko reviewed (approved):
lgtm. Longer term it may be better to try to go back to using the pre-compiled one from debian, or alternatively try a different Linux distro for the msan tasks?

theuni commented:
Concept ACK. I'm not so familiar with the history here, but the changes seem sane to me.

The libc++ changes here are especially interesting. I assume they work with no issues because there's no SDK to take into consideration, but (as @fanquake suggested in an offline discussion yesterday) I'm going to see if we can maybe port this over to our depends builds with the SDK included. Unsure if it'll work 🤷 but it would be much cleaner if it did.

fanquake commented:
> I am not sure if compiling a compiler for a CI task is a worthwhile thing to maintain in this repo? 

This point is in regards to the libc++ flag usage. Not compiling a compiler for anything else.

fanquake commented on ci/test/01_base_install.sh:
```diff
@@ -40,11 +40,32 @@ if [ -n "$PIP_PACKAGES" ]; then
+                                                            -DCMAKE_BUILD_TYPE=Release \
+                                                            -DLLVM_TARGETS_TO_BUILD=X86 \
+                                                            -DLLVM_ENABLE_RUNTIMES="compiler-rt;libcxx;libcxxabi;libunwind" \
+                                                            -S "${BASE_SCRATCH_DIR}"/msan/llvm-project/llvm
```
> Looks like memory usage isn't reported in https://cirrus-ci.com/task/4543816401682432 ?

Is that a CIrrus bug? You should be able to see an example memory usage here: https://cirrus-ci.com/task/4632561431871488.

fanquake replied:
> Looks like it is only passing CI because Cirrus seems to be ignoring the 2CPU/8GB limit and just uses 4/15 unconditionally for ci image builds?

@fkorotkov is this what should be happening, and can we rely on this behaviour into the future?

dergoegge commented:
utACK 5763b232e6e6a0f72d046f8aa322b39328be135b

Looking forward to green check marks on qa-assets

fanquake commented:
Using alternate distros, or the pre-compiled LLVM bins do not currently seem to be viable options. Follow up in the qa-assets repo is here: https://github.com/bitcoin-core/qa-assets/pull/129.

fanquake commented:
> Did this break the llvm symbolizer?

I'll follow up with this.
